### ISSUE TITLE
有流式返回的方法吗？

### ISSUE THREAD

**Issue by czhcc** [OpenLMLab/MOSS#110]
这样在界面上不会停一段时间才看到结果

**Comment by alin995** [comment]
自己封装，把moss_inference.py里的forward, sample方法里的 return 改成 yield


class InferenceExtend(Inference):

    @torch.no_grad()
    def stream_forward(self, data: str, paras: Optional[Dict[str, float]] = None) -> List[str]:
        input_ids, attention_mask = self.preprocess(data)

        if not paras:
            paras = self.default_paras

        for output_ids in self.stream_sample(
                input_ids,
                attention_mask,
                temperature=paras["temperature"],
                repetition_penalty=paras["repetition_penalty"],
                top_k=paras["top_k"],
                top_p=paras["top_p"],
                max_iterations=paras["max_iterations"],
                regulation_start=paras["regulation_start"],
                length_penalty=paras["length_penalty"],
                max_time=paras["max_time"],
        ):
            preds = self.tokenizer.batch_decode(output_ids)
            yield [self.postprocess_remove_prefix(pred) for pred in preds][0]

    def stream_sample(self,
                      input_ids: torch.Tensor,
                      attention_mask: torch.Tensor,
                      temperature: float = 0.7,
                      repetition_penalty: float = 1.1,
                      top_k: int = 0,
                      top_p: float = 0.92,
                      max_iterations: int = 1024,
                      regulation_start: int = 512,
                      length_penalty: float = 1,
                      max_time: int = 60,
                      ) -> torch.Tensor:
        assert input_ids.dtype == torch.int64 and attention_mask.dtype == torch.int64

        self.bsz, self.seqlen = input_ids.shape

        input_ids, attention_mask = input_ids.to('cuda'), attention_mask.to('cuda')
        last_token_indices = attention_mask.sum(1) - 1

        moss_stopwords = self.moss_stopwords.to(input_ids.device)
        queue_for_moss_stopwords = torch.empty(size=(self.bsz, len(self.moss_stopwords)), device=input_ids.device, dtype=input_ids.dtype)
        all_shall_stop = torch.tensor([False] * self.bsz, device=input_ids.device)
        moss_stop = torch.tensor([False] * self.bsz, device=input_ids.device)

        generations, start_time = torch.ones(self.bsz, 1, dtype=torch.int64), time.time()

        past_key_values = None
        new_generated_id = None
        for i in range(int(max_iterations)):
            logits, past_key_values = self.infer_(input_ids if i == 0 else new_generated_id, attention_mask, past_key_values)

            if i == 0:
                logits = logits.gather(1, last_token_indices.view(self.bsz, 1, 1).repeat(1, 1, self.vocab_size)).squeeze(1)
            else:
                logits = logits[:, -1, :]

            if repetition_penalty > 1:
                score = logits.gather(1, input_ids)
                # if score < 0 then repetition penalty has to be multiplied to reduce the previous token probability
                # just gather the histroy token from input_ids, preprocess then scatter back
                # here we apply extra work to exclude special token

                score = torch.where(score < 0, score * repetition_penalty, score / repetition_penalty)

                logits.scatter_(1, input_ids, score)

            logits = logits / temperature

            filtered_logits = self.top_k_top_p_filtering(logits, top_k, top_p)
            probabilities = torch.softmax(filtered_logits, dim=-1)

            cur_len = i
            if cur_len > int(regulation_start):
                for i in self.moss_stopwords:
                    probabilities[:, i] = probabilities[:, i] * pow(length_penalty, cur_len - regulation_start)

            new_generated_id = torch.multinomial(probabilities, 1)

            # update extra_ignored_tokens
            new_generated_id_cpu = new_generated_id.cpu()

            input_ids, attention_mask = torch.cat([input_ids, new_generated_id], dim=1), torch.cat(
                [attention_mask, torch.ones((self.bsz, 1), device=attention_mask.device, dtype=attention_mask.dtype)], dim=1)

            generations = torch.cat([generations, new_generated_id.cpu()], dim=1)

            # stop words components
            queue_for_moss_stopwords = torch.cat([queue_for_moss_stopwords[:, 1:], new_generated_id], dim=1)

            moss_stop |= (queue_for_moss_stopwords == moss_stopwords).all(1)

            all_shall_stop |= moss_stop

            yield input_ids

            if all_shall_stop.all().item():
                break
            elif time.time() - start_time > max_time:
                break

**Comment by czhcc** [comment]
多谢

**Comment by piglaker** [comment]
已解决
